test/common/stats/symbol_table_impl_test.cc

```diff
@@ -263,62 +311,131 @@ TEST_F(StatNameTest, HashTable) {
-  StatNameJoiner joiner(makeStat(""), makeStat("c.d"));
-  EXPECT_EQ("c.d", joiner.statName().toString(table_));
+TEST_P(StatNameTest, ConcatSecondEmpty) {
+  SymbolTable::StoragePtr joined = table_->join({makeStat("a.b"), makeStat("")});
+  EXPECT_EQ("a.b", table_->toString(StatName(joined.get())));
 }
 
-TEST_F(StatNameTest, ConcatSecondEmpty) {
-  StatNameJoiner joiner(makeStat("a.b"), makeStat(""));
-  EXPECT_EQ("a.b", joiner.statName().toString(table_));
+TEST_P(StatNameTest, ConcatAllEmpty) {
+  SymbolTable::StoragePtr joined = table_->join({makeStat(""), makeStat("")});
+  EXPECT_EQ("", table_->toString(StatName(joined.get())));
 }
 
-TEST_F(StatNameTest, ConcatAllEmpty) {
-  StatNameJoiner joiner(makeStat(""), makeStat(""));
-  EXPECT_EQ("", joiner.statName().toString(table_));
+TEST_P(StatNameTest, Join3) {
+  SymbolTable::StoragePtr joined =
+      table_->join({makeStat("a.b"), makeStat("c.d"), makeStat("e.f")});
+  EXPECT_EQ("a.b.c.d.e.f", table_->toString(StatName(joined.get())));
 }
 
-TEST_F(StatNameTest, Join3) {
-  StatNameJoiner joiner({makeStat("a.b"), makeStat("c.d"), makeStat("e.f")});
-  EXPECT_EQ("a.b.c.d.e.f", joiner.statName().toString(table_));
+TEST_P(StatNameTest, Join3FirstEmpty) {
+  SymbolTable::StoragePtr joined = table_->join({makeStat(""), makeStat("c.d"), makeStat("e.f")});
+  EXPECT_EQ("c.d.e.f", table_->toString(StatName(joined.get())));
 }
 
-TEST_F(StatNameTest, Join3FirstEmpty) {
-  StatNameJoiner joiner({makeStat(""), makeStat("c.d"), makeStat("e.f")});
-  EXPECT_EQ("c.d.e.f", joiner.statName().toString(table_));
+TEST_P(StatNameTest, Join3SecondEmpty) {
+  SymbolTable::StoragePtr joined = table_->join({makeStat("a.b"), makeStat(""), makeStat("e.f")});
+  EXPECT_EQ("a.b.e.f", table_->toString(StatName(joined.get())));
 }
 
-TEST_F(StatNameTest, Join3SecondEmpty) {
-  StatNameJoiner joiner({makeStat("a.b"), makeStat(""), makeStat("e.f")});
-  EXPECT_EQ("a.b.e.f", joiner.statName().toString(table_));
+TEST_P(StatNameTest, Join3ThirdEmpty) {
+  SymbolTable::StoragePtr joined = table_->join({makeStat("a.b"), makeStat("c.d"), makeStat("")});
+  EXPECT_EQ("a.b.c.d", table_->toString(StatName(joined.get())));
 }
 
-TEST_F(StatNameTest, Join3ThirdEmpty) {
-  StatNameJoiner joiner({makeStat("a.b"), makeStat("c.d"), makeStat("")});
-  EXPECT_EQ("a.b.c.d", joiner.statName().toString(table_));
+TEST_P(StatNameTest, JoinAllEmpty) {
+  SymbolTable::StoragePtr joined = table_->join({makeStat(""), makeStat(""), makeStat("")});
+  EXPECT_EQ("", table_->toString(StatName(joined.get())));
 }
 
-TEST_F(StatNameTest, JoinAllEmpty) {
-  StatNameJoiner joiner({makeStat(""), makeStat(""), makeStat("")});
-  EXPECT_EQ("", joiner.statName().toString(table_));
+TEST_P(StatNameTest, RacingSymbolCreation) {
+  Thread::ThreadFactory& thread_factory = Thread::threadFactoryForTest();
+  MutexTracerImpl& mutex_tracer = MutexTracerImpl::getOrCreateTracer();
+
+  // Make 100 threads, each of which will race to encode an overlapping set of
```

Comment:
Not a huge fan of non-deterministic tests like this, since it's hard to get a repeater. Has this been historically useful? Do you use TSAN to get some race info to debug? How hard is it to repro when this happens?

Reply:
Per f2f I have changed this test to actually test something useful: that we get zero contentions when fake symbol tables are employed.